internal/controller/component_build_controller_service_account.go

```diff
@@ -129,6 +139,140 @@ func (r *ComponentBuildReconciler) linkCommonSecretsToBuildPipelineServiceAccoun
+	buildPipelineServiceAccountName := getBuildPipelineServiceAccountName(component)
+	log := ctrllog.FromContext(ctx).WithValues("ServiceAccountName", buildPipelineServiceAccountName)
+
+	allComponentsInNamespaceList := &appstudiov1alpha1.ComponentList{}
+	if err := r.Client.List(ctx, allComponentsInNamespaceList, client.InNamespace(component.Namespace)); err != nil {
+		log.Error(err, "failed to list all Components in namespace", l.Action, l.ActionView)
+		return err
+	}
+
+	// TODO use Component.Status.BuildNudgedBy when KONFLUX-6841 is resolved and the data in status will be reliable.
+	// Find all Components that nudge current one.
+	nudgingComponents := []string{}
+	for _, c := range allComponentsInNamespaceList.Items {
+		for _, componentNameToNudge := range c.Spec.BuildNudgesRef {
+			if componentNameToNudge == component.Name {
+				nudgingComponents = append(nudgingComponents, c.Name)
+				break
+			}
+		}
+	}
+
+	if len(nudgingComponents) == 0 {
+		// No components that nudge current one, nothing to link.
+		return nil
+	}
+
+	buildPipelineServiceAccount := &corev1.ServiceAccount{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: buildPipelineServiceAccountName, Namespace: component.Namespace}, buildPipelineServiceAccount); err != nil {
+		log.Error(err, "failed to get build pipeline Service Account", l.Action, l.ActionView)
+		return err
+	}
+
+	isServiceAccountUpdated := false
+	for _, componentName := range nudgingComponents {
+		imageRepository, err := r.getComponentImageRepository(ctx, componentName, component.Namespace)
+		if err != nil {
+			log.Error(err, "failed to get Image Repository for Component", "Component", componentName, l.Action, l.ActionView)
+			return err
+		}
+		if imageRepository == nil {
+			// The Component doesn't have related Image Repository object.
+			continue
+		}
+		pullSecretName := imageRepository.Status.Credentials.PullSecretName
+		if !isSaSecretLinked(buildPipelineServiceAccount, pullSecretName, false) {
+			buildPipelineServiceAccount.Secrets = append(buildPipelineServiceAccount.Secrets,
+				corev1.ObjectReference{Name: pullSecretName, Namespace: buildPipelineServiceAccount.Namespace})
+			isServiceAccountUpdated = true
+		}
+	}
+	if isServiceAccountUpdated {
+		if err := r.Client.Update(ctx, buildPipelineServiceAccount); err != nil {
+			log.Error(err, "failed to update build pipeline Service Account", l.Action, l.ActionUpdate)
+			return err
+		}
+		log.Info("Updated Service Account pull secrets")
+	}
+
+	return nil
+}
+
+// clenaUpNudgingPullSecrets removes the current Component pull secret from the linked secrets list
+// of the nudged Component build pipeline Service Account.
+func (r *ComponentBuildReconciler) clenaUpNudgingPullSecrets(ctx context.Context, component *appstudiov1alpha1.Component) error {
+	log := ctrllog.FromContext(ctx).WithName("cleanupNudgingPullSecrets")
+
```

Comment:
check 1st if component.Spec.BuildNudgesRef is empty and if it is return nil

Reply:
good point